app/jobs/autoscaling.py

```diff
@@ -23,14 +25,44 @@
 SCALING_METRIC_TAG_PREFIX = "IDSeqScalingData_"
 SCALING_METRIC_TAG_FORMAT = SCALING_METRIC_TAG_PREFIX + "{environment}"
 
-
 # A value recorded this many minutes earlier will be ignored
 EXPIRATION_PERIOD_MINUTES = 60
 
-
 DEBUG = False
 
 
+## constants specific to draining / safe scale-down:
+GRACE_PERIOD_SECONDS__JOB_DISPATCH_LAG = 300
+GRACE_PERIOD_SECONDS__JOB_TAG_KEEP_ALIVE = 300
+
+
+## EXPLANATION OF STATE TRANSITIONS FOR GSNAP/RAPSEARCH MACHINES
+'''
+{ service: gsnap-prod-initializing, ProtectedFromScaleIn: True }
+Initializing. Scale-in protection automatically granted at birth.
+    |
+    V
+{ service: gsnap-prod, ProtectedFromScaleIn: True }
+In service, accepting chunks. Scale-in protection present.
+    |
+    V
+{ service: gsnap-prod, ProtectedFromScaleIn: True,
+  RunningIDseqBatchJob_X1: Y1, RunningIDseqBatchJob_X2: Y2, ...
+}
+Processing chunks.
+    |
+    V
+{ service: gsnap-prod, draining: T0, ProtectedFromScaleIn: True }
+Draining. Scale-in protection still present.
+Can still accept chunks for up to min_draining_wait seconds after entering this state.
+Can still finish running chunks for up to (min_draining_wait + job_tag_expiration) seconds after entering this state.
+    |
+    V
+{ service: gsnap-prod, draining: T0, ProtectedFromScaleIn: False }
+Discarded. Scale-in protection withdrawn on transition to this state.
+'''
```

Comment:
Thanks for documenting these states and transitions.    Two of the states in the middle seem identical, should be compressed to single state.   You are missing the back arrow for reclaiming instances from draining.  What is the meaning of T0?

In computer science the word "state" is fairly overloaded, but so far as state machines, it's pretty standard to define a state machine as a set of states and set of transitions;   you can label those states S1, S2, ... and list in a table explaining what they are (like you have above) but then list the transitions S1 -> S2, S2 -> S3, S3 -> S4, and S3 -> S2 in a separate table perhaps.  Each transition also needs a word of explanation.